README.md
# WordSimilarity
Calc Similarity between two english words

Implementation of my blog(In Chinese): http://www.cnblogs.com/XBWer/p/3854440.html.


JWordNetSim/HISTORY
Java WordNet Similarity v1.0.0
Copyright (C) 2006-2007 The University of Sheffield
Developed by Mark A. Greenwood <m.greenwood@dcs.shef.ac.uk>

=======================================================================
Version 1.0.0 (Released 13/04/2007)
=======================================================================
The initial public release. This currently only contains the JCn and
Lin measures and supporting code. Hopefully more measures will be added
very soon.

JWordNetSim/test/Test.java
/************************************************************************
 *         Copyright (C) 2006-2007 The University of Sheffield          *
 *      Developed by Mark A. Greenwood <m.greenwood@dcs.shef.ac.uk>     *
 *                                                                      *
 * This program is free software; you can redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the Free Software Foundation; either version 2 of the License, or    *
 * (at your option) any later version.                                  *
 *                                                                      *
 * This program is distributed in the hope that it will be useful,      *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have received a copy of the GNU General Public License    *
 * along with this program; if not, write to the Free Software          *
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 ************************************************************************/
package JWordNetSim.test;

import java.io.FileInputStream;
import java.util.HashMap;
import java.util.Map;

import net.didion.jwnl.JWNL;
import net.didion.jwnl.data.IndexWord;
import net.didion.jwnl.data.POS;
import net.didion.jwnl.dictionary.Dictionary;
import shef.nlp.wordnet.similarity.SimilarityMeasure;

/**
 * A simple test of this WordNet similarity library.
 * @author Mark A. Greenwood
 */
public class Test
{
	public static void main(String[] args) throws Exception
	{	
		//Initialize WordNet - this must be done before you try
		//and create a similarity measure otherwise nasty things
		//might happen!
		JWNL.initialize(new FileInputStream("D:\\JAVAProjectWorkSpace\\jwnl\\JWordNetSim\\test\\wordnet.xml"));
		
		//Create a map to hold the similarity config params
		Map<String,String> params = new HashMap<String,String>();
		
		//the simType parameter is the class name of the measure to use
		params.put("simType","shef.nlp.wordnet.similarity.JCn");
		
		//this param should be the URL to an infocontent file (if required
		//by the similarity measure being loaded)
		params.put("infocontent","file:D:\\JAVAProjectWorkSpace\\jwnl\\JWordNetSim\\test\\ic-bnc-resnik-add1.dat");
		
		//this param should be the URL to a mapping file if the
		//user needs to make synset mappings
		params.put("mapping","file:D:\\JAVAProjectWorkSpace\\jwnl\\JWordNetSim\\test\\domain_independent.txt");
		
		//create the similarity measure
		SimilarityMeasure sim = SimilarityMeasure.newInstance(params);
		
		//Get two words from WordNet
		Dictionary dict = Dictionary.getInstance();		
		IndexWord word1 = dict.getIndexWord(POS.NOUN, "dog");
		IndexWord word2 = dict.getIndexWord(POS.NOUN,"cat");
		
		//and get the similarity between the first senses of each word
		System.out.println(word1.getLemma()+"#"+word1.getPOS().getKey()+"#1  " + word2.getLemma()+"#"+word2.getPOS().getKey()+"#1  " + sim.getSimilarity(word1.getSense(1), word2.getSense(1)));		
		
		//get similarity using the string methods (note this also makes use
		//of the fake root node)
		System.out.println(sim.getSimilarity("cat#n","cat#n"));
		
		//get a similarity that involves a mapping
		System.out.println(sim.getSimilarity("namperson", "namperson"));
	}
}


shef/nlp/wordnet/similarity/ICMeasure.java
/************************************************************************
 *         Copyright (C) 2006-2007 The University of Sheffield          *
 *      Developed by Mark A. Greenwood <m.greenwood@dcs.shef.ac.uk>     *
 *                                                                      *
 * This program is free software; you can redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the Free Software Foundation; either version 2 of the License, or    *
 * (at your option) any later version.                                  *
 *                                                                      *
 * This program is distributed in the hope that it will be useful,      *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have received a copy of the GNU General Public License    *
 * along with this program; if not, write to the Free Software          *
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 ************************************************************************/

package shef.nlp.wordnet.similarity;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import net.didion.jwnl.JWNL;
import net.didion.jwnl.JWNLException;
import net.didion.jwnl.data.POS;
import net.didion.jwnl.data.Pointer;
import net.didion.jwnl.data.PointerUtils;
import net.didion.jwnl.data.Synset;
import net.didion.jwnl.data.Word;
import net.didion.jwnl.data.list.PointerTargetNode;

/**
 * An abstract class that addes information content based methods to the
 * top level similarity measure class but doesn't itself define a
 * similarity measure.
 * @author Mark A. Greenwood
 */
public abstract class ICMeasure extends PathMeasure
{
	/**
	 * This map stores the synset IDs and there associated frequencies
	 * as read from the supplied information content file.
	 */
	private Map<String,Double> freq = new HashMap<String,Double>();
			
	protected void config(Map<String,String> params) throws Exception
	{
		super.config(params);
		
		//a handle to the infocontent file
		BufferedReader in = null;
		
		try
		{
			URL url = new URL(params.remove("infocontent"));
			
			//open the info content file for reading
			in = new BufferedReader(new InputStreamReader(url.openStream()));
			
			//get the first line from the file (should be the WordNet version info)
			String line = in.readLine();
					
			//Check that what we have is actually a file of IC values
			if (line == null || !line.startsWith("wnver::")) throw new IOException("Malformed InfoContent file");
			
			//Check that the IC file is meant for use with the version
			//of WordNet we are currently using
			if (!line.endsWith("::"+JWNL.getVersion().getNumber())) throw new Exception("InfoContent file version doesn't match WordNet version");
			
			//Initially set the IC values of the noun and verb roots to 0
			freq.put("n",0d);
			freq.put("v",0d);
			
			//Get the first line of real data ready for use
			line = in.readLine();		
			
			while (line != null && !line.equals(""))
			{
				//while there is still data in the file to process...
				
				//split the line on the whitespace
				String[] data = line.split("\\s+");
				
				//store the frequency (2nd column) against the synset ID (1st column)
				freq.put(data[0],new Double(data[1]));
				
				if (data.length == 3 && data[2].equals("ROOT"))
				{
					//if there are three columns on this line and the
					//last one is ROOT then...
					
					//get the POS tag of the synset
					String pos = data[0].substring(data[0].length()-1);
					
					//updated the node frequency for the POS tag
					freq.put(pos, Double.parseDouble(data[1])+freq.get(pos));
				}
				
				//read in the next line from the file ready for processing
				line = in.readLine();
			}
		}
		finally
		{
			//if we managed to open the file then close it
			if (in != null) in.close();
		}
	}
	
	/**
	 * Generates the key to access the frequency count data loaded
	 * from the information content file.
	 * @param synset the synset for which to generate the key.
	 * @return the key to access the frequency count map.
	 */
	protected String getFreqKey(Synset synset)
	{
		//the keys used by the infomation content files are simply
		//the offsets in the wordnet database (minus leading zeros)
		//followed by the single character POS tag. So simply build
		//a key of this type...
				
		return synset.getOffset()+synset.getPOS().getKey();
	}
	
	/**
	 * Gets the Information Content (IC) value associated with the given synset.
	 * @param synset the synset for which to calcualte IC.
	 * @return the IC of the given synset.
	 */
	protected double getIC(Synset synset)
	{
		//get the POS tag of this synset
		POS pos = synset.getPOS();
		
		//Information Content is only defined for nouns and verbs
		//so return 0 if the POS tag is something else
		if (!pos.equals(POS.NOUN) && !pos.equals(POS.VERB)) return 0;
		
		//Get the frequency of this synset from the storred data
		Double synFreq = freq.get(getFreqKey(synset));
				
		//if the frequency isn't defined or it's 0 then simlpy return 0 
		if (synFreq == null || synFreq.doubleValue() == 0) return 0;
		
		//Get the frequency of the root node for this POS tage
		Double rootFreq = freq.get(synset.getPOS().getKey());
	
		//calcualte the probability for this synset
		double prob = synFreq.doubleValue() / rootFreq.doubleValue();
		
		//if the probability is valid then use it to return the IC value
		if (prob > 0) return -Math.log(prob);
		
		//something went wrong so assume IC of 0
		return 0;
	}
	
	/**
	 * Returns the frequency of the root node of the hierarchy for the
	 * given POS tag.
	 * @param pos the POS tag of the root node to access
	 * @return the frequency of the root node for the given POS tag
	 */
	protected double getFrequency(POS pos)
	{
		return freq.get(pos.getKey());
	}
	
	/**
	 * Returns the frequency of the given synset.
	 * @param synset the synset to retrieve the frequency of
	 * @return the frequency of the supplied synset
	 */
	protected double getFrequency(Synset synset)
	{
		Double f = freq.get(getFreqKey(synset));
		
		if (f == null || f.doubleValue() == 0) return 0;
		
		return f.doubleValue();
	}
	
	/**
	 * Finds the lowerst common subsumer of the two synsets using information content.
	 * @param s1 the first synset
	 * @param s2 the second synset
	 * @return the lowest common subsumer of the two provided synsets
	 * @throws JWNLException if an error occurs accessing WordNet
	 */
	protected Synset getLCSbyIC(Synset s1, Synset s2) throws JWNLException
	{
		//TODO Handle the different types of LCS handled by the perl version which are
		//   1) Largest IC value
		//   2) Results in shortest path
		//   3) Greatest depth (i.e. the LCS whose shortest path to root is longest)
		//Although in here we only need the IC based one
				
		@SuppressWarnings("unchecked")
		List<List<PointerTargetNode>> trees1 = PointerUtils.getInstance().getHypernymTree(s1).toList();
		
		@SuppressWarnings("unchecked")
		List<List<PointerTargetNode>> trees2 = PointerUtils.getInstance().getHypernymTree(s2).toList();
		
		Set<Synset> pLCS = new HashSet<Synset>();
		
		for (List<PointerTargetNode> t1 : trees1)
		{
			for (List<PointerTargetNode> t2 : trees2)
			{
				for (PointerTargetNode node : t1)
				{
					if (contains(t2,node.getSynset()))
					{
						pLCS.add(node.getSynset());
						break;
					}
				}
				
				for (PointerTargetNode node : t2)
				{
					if (contains(t1,node.getSynset()))
					{
						pLCS.add(node.getSynset());
						break;
					}
				}
			}
		}
		
		Synset lcs = null;
		double score = 0;
		
		for (Synset s : pLCS)
		{
			if (lcs == null)
			{
				lcs = s;
				score = getIC(s);
			}
			else
			{
				double ic = getIC(s);
				
				if (ic > score)
				{
					score = ic;
					lcs = s;
				}
			}
		}
		
		
		if (lcs == null && useSingleRoot())
		{	
			//link the two synsets by a fake root node
			
			//TODO: Should probably create one of these for each POS tag and cache them so that we can always return the same one
			lcs = new Synset(s1.getPOS(),0l,new Word[0],new Pointer[0],"",new java.util.BitSet());
						
		}
		
		return lcs;
	}
}


shef/nlp/wordnet/similarity/PathMeasure.java
/************************************************************************
 *         Copyright (C) 2006-2007 The University of Sheffield          *
 *      Developed by Mark A. Greenwood <m.greenwood@dcs.shef.ac.uk>     *
 *                                                                      *
 * This program is free software; you can redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the Free Software Foundation; either version 2 of the License, or    *
 * (at your option) any later version.                                  *
 *                                                                      *
 * This program is distributed in the hope that it will be useful,      *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have received a copy of the GNU General Public License    *
 * along with this program; if not, write to the Free Software          *
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 ************************************************************************/

package shef.nlp.wordnet.similarity;

import java.util.List;
import java.util.Map;

import net.didion.jwnl.data.Synset;
import net.didion.jwnl.data.list.PointerTargetNode;

/**
 * An abstract class that addes path based methods to the top level similarity
 * measure class but doesn't itself define a similarity measure.
 * @author Mark A. Greenwood
 */
public abstract class PathMeasure extends SimilarityMeasure
{
	/**
	 * If true then a fake root node is used to joing the multiple
	 * verb and noun hierarchies into one hierarchy per POS tag
	 */
	private boolean root = true;
	
	/**
	 * Should we use a siingle root node for each POS tag hierarchy
	 * @return true if we should use a single root node for each POS tag, false otherwise
	 */
	protected boolean useSingleRoot() { return root; }
	
	protected void config(Map<String,String> params) throws Exception
	{
		//A protected constructor to force the use of the newInstance method
		
		if (params.containsKey("root")) root = Boolean.parseBoolean(params.remove("root"));
	}
	
	/**
	 * Utility method to determine if the list of nodes contains a given synset
	 * @param l a list of nodes
	 * @param s a synset
	 * @return true if the synset is contained within the list of nodes, false otherwise
	 */
	protected static boolean contains(List<PointerTargetNode> l, Synset s)
	{
		for (PointerTargetNode node : l)
		{
			if (node.getSynset().equals(s)) return true;
		}
		
		return false;
	}
}


JWordNetSim/src/shef/nlp/wordnet/similarity/JCn.java
/************************************************************************
 *         Copyright (C) 2006-2007 The University of Sheffield          *
 *      Developed by Mark A. Greenwood <m.greenwood@dcs.shef.ac.uk>     *
 *                                                                      *
 * This program is free software; you can redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the Free Software Foundation; either version 2 of the License, or    *
 * (at your option) any later version.                                  *
 *                                                                      *
 * This program is distributed in the hope that it will be useful,      *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have received a copy of the GNU General Public License    *
 * along with this program; if not, write to the Free Software          *
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 ************************************************************************/

package shef.nlp.wordnet.similarity;

import net.didion.jwnl.JWNLException;
import net.didion.jwnl.data.Synset;

/**
 * An implementation of the WordNet similarity measure developed by Jiang and
 * Conrath. For full details of the measure see:
 * <blockquote>Jiang J. and Conrath D. 1997. Semantic similarity based on corpus
 * statistics and lexical taxonomy. In Proceedings of International
 * Conference on Research in Computational Linguistics, Taiwan.</blockquote>
 * @author Mark A. Greenwood
 */
public class JCn extends ICMeasure
{
	/**
	 * Instances of this similarity measure should be generated using the
	 * factory methods of {@link SimilarityMeasure}.
	 */
	protected JCn()
	{
		//A protected constructor to force the use of the newInstance method
	}
	
	@Override public double getSimilarity(Synset s1, Synset s2) throws JWNLException
	{
		//if the POS tags are not the same then return 0 as this measure
		//only works with 2 nouns or 2 verbs.
		if (!s1.getPOS().equals(s2.getPOS())) return 0;
		
		//see if the similarity is already cached and...
		Double cached = getFromCache(s1, s2);
		
		//if it is then simply return it
		if (cached != null) return cached.doubleValue();
		
		//Get the Information Content (IC) values for the two supplied synsets
		double ic1 = getIC(s1);
		double ic2 = getIC(s2);

		//if either IC value is zero then cache and return a sim of 0
		if (ic1 == 0 || ic2 == 0) return addToCache(s1,s2,0);
		
		//Get the Lowest Common Subsumer (LCS) of the two synsets
		Synset lcs = getLCSbyIC(s1,s2);
		
		//if there isn't an LCS then cache and return a sim of 0
		if (lcs == null) return addToCache(s1,s2,0);
		
		//get the IC valueof the LCS
		double icLCS = getIC(lcs);
		
		//compute the distance between the two synsets
		//NOTE: This is the original JCN measure
		double distance = ic1 + ic2 - (2 * icLCS);
		
		//assume the similarity between the synsets is 0
		double sim = 0;
		
		if (distance == 0)
		{
			//if the distance is 0 (i.e. ic1 + ic2 = 2 * icLCS) then...
			
			//get the root frequency for this POS tag
			double rootFreq = getFrequency(s1.getPOS());
			
			if (rootFreq > 0.01)
			{
				//if the root frequency has a value then use it to generate a
				//very large sim value
				sim = 1/-Math.log((rootFreq - 0.01) / rootFreq);
		    }		    
		}
		else
		{
			//this is the normal case so just convert the distance
			//to a similarity by taking the multiplicative inverse
			sim = 1/distance;
		}
		
		//cache and return the calculated similarity
		return addToCache(s1,s2,sim);
	}
}


JWordNetSim/src/shef/nlp/wordnet/similarity/Lin.java
/************************************************************************
 *         Copyright (C) 2006-2007 The University of Sheffield          *
 *      Developed by Mark A. Greenwood <m.greenwood@dcs.shef.ac.uk>     *
 *                                                                      *
 * This program is free software; you can redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the Free Software Foundation; either version 2 of the License, or    *
 * (at your option) any later version.                                  *
 *                                                                      *
 * This program is distributed in the hope that it will be useful,      *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have received a copy of the GNU General Public License    *
 * along with this program; if not, write to the Free Software          *
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 ************************************************************************/

package shef.nlp.wordnet.similarity;

import net.didion.jwnl.JWNLException;
import net.didion.jwnl.data.Synset;

/**
 * An implementation of the WordNet similarity measure developed by Lin. For
 * full details of the measure see:
 * <blockquote>Lin D. 1998. An information-theoretic definition of similarity. In
 * Proceedings of the 15th International Conference on Machine
 * Learning, Madison, WI.</blockquote>
 * @author Mark A. Greenwood
 */
public class Lin extends ICMeasure
{
	/**
	 * Instances of this similarity measure should be generated using the
	 * factory methods of {@link SimilarityMeasure}.
	 */
	protected Lin()
	{
		//A protected constructor to force the use of the newInstance method
	}
	
	@Override public double getSimilarity(Synset s1, Synset s2) throws JWNLException
	{
		//if the POS tags are not the same then return 0 as this measure
		//only works with 2 nouns or 2 verbs.
		if (!s1.getPOS().equals(s2.getPOS())) return 0;
		
		//see if the similarity is already cached and...
		Double cached = getFromCache(s1, s2);
		
		//if it is then simply return it
		if (cached != null) return cached.doubleValue();
		
		//Get the Information Content (IC) values for the two supplied synsets
		double ic1 = getIC(s1);
		double ic2 = getIC(s2);
		
		//if either IC value is zero then cache and return a sim of 0
		if (ic1 == 0 || ic2 == 0) return addToCache(s1,s2,0);
		
		//Get the Lowest Common Subsumer (LCS) of the two synsets
		Synset lcs = getLCSbyIC(s1,s2);
		
		//if there isn't an LCS then cache and return a sim of 0
		if (lcs == null) return addToCache(s1,s2,0);
		
		//get the IC valueof the LCS
		double icLCS = getIC(lcs);
		
		//caluclaue the similarity score
		double sim = (2*icLCS)/(ic1+ic2);
		
		//cache and return the calculated similarity
		return addToCache(s1,s2,sim);
	}
}


JWordNetSim/src/shef/nlp/wordnet/similarity/SimilarityMeasure.java
/************************************************************************
 *         Copyright (C) 2006-2007 The University of Sheffield          *
 *      Developed by Mark A. Greenwood <m.greenwood@dcs.shef.ac.uk>     *
 *                                                                      *
 * This program is free software; you can redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the Free Software Foundation; either version 2 of the License, or    *
 * (at your option) any later version.                                  *
 *                                                                      *
 * This program is distributed in the hope that it will be useful,      *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have received a copy of the GNU General Public License    *
 * along with this program; if not, write to the Free Software          *
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 ************************************************************************/

package shef.nlp.wordnet.similarity;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

import net.didion.jwnl.JWNLException;
import net.didion.jwnl.data.IndexWord;
import net.didion.jwnl.data.POS;
import net.didion.jwnl.data.Synset;
import net.didion.jwnl.dictionary.Dictionary;

/**
 * An abstract notion of a similarity measure that all provided
 * implementations extend.
 * @author Mark A. Greenwood
 */
public abstract class SimilarityMeasure
{	
	/**
	 * A mapping of terms to specific synsets. Usually used to map domain
	 * terms to a restricted set of synsets but can also be used to map
	 * named entity tags to appropriate synsets.
	 */
	private Map<String,Set<Synset>> domainMappings = new HashMap<String,Set<Synset>>();
	
	/**
	 * The maximum size the cache can grow to
	 */
	private int cacheSize = 5000;
	
	/**
	 * To speed up computation of the similarity between two synsets
	 * we cache each similarity that is computed so we only have to
	 * do each one once.
	 */
	private Map<String,Double> cache = new LinkedHashMap<String,Double>(16,0.75f,true)
	{
        public boolean removeEldestEntry(Map.Entry<String,Double> eldest)
        {
            //if the size is less than zero then the user is asking us
        	//not to limit the size of the cache so return false
        	if (cacheSize < 0) return false;
        	
        	//if the cache has crown bigger than it's max size return true
        	return size() > cacheSize;
        }
    }; 
	
	/**
	 * Get a previously computed similarity between two synsets from the cache.
	 * @param s1 the first synset between which we are looking for the similarity.
	 * @param s2 the other synset between which we are looking for the similarity.
	 * @return The similarity between the two sets or null
	 *         if it is not in the cache.
	 */
	protected final Double getFromCache(Synset s1, Synset s2)
	{
		return cache.get(s1.getKey()+"-"+s2.getKey());
	}
	
	/**
	 * Add a computed similarity between two synsets to the cache so that
	 * we don't have to compute it if it is needed in the future.
	 * @param s1 one of the synsets between which we are storring a similarity.
	 * @param s2 the other synset between which we are storring a similarity.
	 * @param sim the similarity between the two supplied synsets.
	 * @return the similarity score just added to the cache.
	 */
	protected final double addToCache(Synset s1, Synset s2, double sim)
	{
		cache.put(s1.getKey()+"-"+s2.getKey(),sim);
		
		return sim;
	}
	
	/**
	 * Configures the similarity measure using the supplied parameters.
	 * @param params a set of key-value pairs that are used to configure
	 *        the similarity measure. See concrete implementations for details
	 *        of expected/possible parameters. 
	 * @throws Exception if an error occurs while configuring the similarity measure.
	 */
	protected abstract void config(Map<String,String> params) throws Exception;
	
	/**
	 * Create a new instance of a similarity measure.
	 * @param confURL the URL of a configuration file. Parameters are specified
	 *        one per line as key:value pairs.
	 * @return a new instance of a similairy measure as defined by the
	 *         supplied configuration URL.
	 * @throws Exception if an error occurs while creating the similarity measure.
	 */
	public static SimilarityMeasure newInstance(URL confURL) throws Exception
	{
		//create map to hold the key-value pairs we are going to read from
		//the configuration file
		Map<String,String> params = new HashMap<String,String>();
		
		//create a reader for the config file
		BufferedReader in = null;
		
		try
		{
			//open the config file
			in = new BufferedReader(new InputStreamReader(confURL.openStream()));
					
			String line = in.readLine();
			while (line != null)
			{
				line = line.trim();
				
				if (!line.equals(""))
				{
					//if the line contains something then
					
					//split the data so we get the key and value
					String[] data = line.split("\\s*:\\s*",2);
					
					if (data.length == 2)
					{
						//if the line is valid add the two parts to the map
						params.put(data[0], data[1]);
					}
					else
					{
						//if the line isn't valid tell the user but continue on
						//with the rest of the file
						System.out.println("Config Line is Malformed: " + line);
					}
				}
				
				//get the next line ready to process
				line = in.readLine();
			}
		}
		finally
		{
			//close the config file if it got opened
			if (in != null) in.close();
		}
		
		//create and return a new instance of the similarity measure specified
		//by the config file
		return newInstance(params);
	}
	
	/**
	 * Creates a new instance of a similarity measure using the supplied parameters.
	 * @param params a set of key-value pairs which define the similarity measure.
	 * @return the newly created similarity measure.
	 * @throws Exception if an error occurs  while creating the similarity measure.
	 */
	public static SimilarityMeasure newInstance(Map<String,String> params) throws Exception
	{
		//get the class name of the implementation we need to load
		String name = params.remove("simType");
		
		//if the name hasn't been specified then throw an exception
		if (name == null) throw new Exception("Must specifiy the similarity measure to use");
		
		//Get hold of the class we need to load
		@SuppressWarnings("unchecked") Class<SimilarityMeasure> c = (Class<SimilarityMeasure>)Class.forName(name);
		
		//create a new instance of the similarity measure
		SimilarityMeasure sim = c.newInstance();
		
		//get the cache parameter from the config params
		String cSize = params.remove("cache");
		
		//if a cache size was specified then set it
		if (cSize != null) sim.cacheSize = Integer.parseInt(cSize);
		
		//get the url of the domain mapping file
		String mapURL = params.remove("mapping");
		
		if (mapURL != null)
		{
			//if a mapping file has been provided then 
						
			//open a reader over the file
			BufferedReader in = new BufferedReader(new InputStreamReader((new URL(mapURL)).openStream()));
			
			//get the first line ready for processing
			String line = in.readLine();
			
			while (line != null)
			{
				if (!line.startsWith("#"))
				{
					//if the line isn't a comment (i.e. it doesn't start with #) then...
					
					//split the line at the white space
					String[] data = line.trim().split("\\s+");
					
					//create a new set to hold the mapped synsets
					Set<Synset> mappedTo = new HashSet<Synset>();
					
					for (int i = 1 ; i < data.length ; ++i)
					{
						//for each synset mapped to get the actual Synsets
						//and store them in the set
						mappedTo.addAll(sim.getSynsets(data[i]));
					}
					
					//if we have found some actual synsets then
					//store them in the domain mappings
					if (mappedTo.size() > 0) sim.domainMappings.put(data[0], mappedTo);
				}
				
				//get the next line from the file
				line = in.readLine();
			}
			
			//we have finished with the mappings file so close it
			in.close();
		}		
		
		//make sure it is configured properly
		sim.config(params);
		
		//then return it
		return sim;
	}
	
	/**
	 * This is the method responsible for computing the similarity between two
	 * specific synsets. The method is implemented differently for each
	 * similarity measure so see the subclasses for detailed information.
	 * @param s1 one of the synsets between which we want to know the similarity.
	 * @param s2 the other synset between which we want to know the similarity.
	 * @return the similarity between the two synsets.
	 * @throws JWNLException if an error occurs accessing WordNet.
	 */
	public abstract double getSimilarity(Synset s1, Synset s2) throws JWNLException;
	
	/**
	 * Get the similarity between two words. The words can be specified either
	 * as just the word or in an encoded form including the POS tag and possibly
	 * the sense number, i.e. cat#n#1 would specifiy the 1st sense of the noun cat.
	 * @param w1 one of the words to compute similarity between.
	 * @param w2 the other word to compute similarity between.
	 * @return a SimilarityInfo instance detailing the similarity between the
	 *         two words specified.
	 * @throws JWNLException if an error occurs accessing WordNet.
	 */
	public final SimilarityInfo getSimilarity(String w1, String w2) throws JWNLException
	{
		//Get the (possibly) multiple synsets associated with each word
		Set<Synset> ss1 = getSynsets(w1);
		Set<Synset> ss2 = getSynsets(w2);
				
		//assume the words are not at all similar
		SimilarityInfo sim = null;
		
		for (Synset s1 : ss1)
		{
			for (Synset s2 : ss2)
			{
				//for each pair of synsets get the similarity
				double score = getSimilarity(s1, s2);
								
				if (sim == null || score > sim.getSimilarity())
				{
					//if the similarity is better than we have seen before
					//then create and store an info object describing the
					//similarity between the two synsets
					sim = new SimilarityInfo(w1, s1, w2, s2, score);
				}
			}
		}
		
		//return the maximum similarity we have found
		return sim;	
	}
	
	/**
	 * Finds all the synsets associated with a specific word.
	 * @param word the word we are interested. Note that this may be encoded
	 *        to include information on POS tag and sense index.
	 * @return a set of synsets that are associated with the supplied word
	 * @throws JWNLException if an error occurs accessing WordNet
	 */
	private final Set<Synset> getSynsets(String word) throws JWNLException
	{		
		//get a handle on the WordNet dictionary
		Dictionary dict = Dictionary.getInstance();
		
		//create an emptuy set to hold any synsets we find
		Set<Synset> synsets = new HashSet<Synset>();
		
		//split the word on the # characters so we can get at the
		//upto three componets that could be present: word, POS tag, sense index
		String[] data = word.split("#");
		
		//if the word is in the domainMappings then simply return the mappings
		if (domainMappings.containsKey(data[0])) return domainMappings.get(data[0]);
		
		if (data.length == 1)
		{
			//if there is just the word
				
			for (IndexWord iw : dict.lookupAllIndexWords(data[0]).getIndexWordArray())
			{
				//for each matching word in WordNet add all it's senses to
				//the set we are building up
				synsets.addAll(Arrays.asList(iw.getSenses()));
			}
			
			//we have finihsed so return the synsets we found
			return synsets;
		}
	
		//the calling method specified a POS tag as well so get that
		POS pos = POS.getPOSForKey(data[1]);
		
		//if the POS tag isn't valid throw an exception
		if (pos == null) throw new JWNLException("Invalid POS Tag: " + data[1]);
		
		//get the word with the specified POS tag from WordNet
		IndexWord iw = dict.getIndexWord(pos, data[0]);
		
		if (data.length > 2)
		{
			//if the calling method specified a sense index then
			//add just that sysnet to the set we are creating
			synsets.add(iw.getSense(Integer.parseInt(data[2])));
		}
		else
		{
			//no sense index was specified so add all the senses of
			//the word to the set we are creating
			synsets.addAll(Arrays.asList(iw.getSenses()));
		}
		
		//return the set of synsets we found for the specified word
		return synsets;
	}
}
